crates/llama-cpp-bindings/src/engine.cc

```diff
@@ -111,7 +134,10 @@ size_t weighted_random(const float* nums, size_t len, uint64_t seed) {
   for (i = 0; i < len; i++) {
     sum += nums[i];
     if (sum >= random) {
-      return i;
+      const auto token_str = llama_token_to_piece(ctx, i);
+      if (req->check_candidate(token_str)) {
```

Comment:
@boxbeam could you help confirm if this approach is making sense from distribution perspective?

Reply:
No, I don't think this makes sense. This means that weighted_random will keep going and select the next token after this one if the check fails. So if the first token selected doesn't work, it'll select whatever the next one is, regardless of its weight, as long as it passes `check_candidate`. We most likely want to perform this logic outside `weighted_random`, repeatedly calling it and checking the output token until we find a token we like.

Reply:
offline discussed